### ISSUE TITLE
"Cannot allocate memory" when creating child process

### ISSUE THREAD

**Issue by AlexEzzeddine** [jruby/jruby#5911]
### Environment

jruby version: 9.2.8.0
ubuntu 14.04

### Expected Behavior

When creating a child process from jruby, the child process should not copy the whole memory of the parent process and should probably use copy-on-write mode of forking (`vfork` instead of `fork` if I'm not mistaken).

### Actual Behavior

We are running a large application (8 gb of memory) on a box with 16gb total memory. When our app calls external program (small memory usage), the child process tries to allocate 8 more GBs of memory and therefore crashes with `Errno::ENOMEM: Cannot allocate memory`.

In other words, child process copies the whole memory of the parent process when executing `fork` system call.

Here is a small snippet:
https://gist.github.com/AlexEzzeddine/5b493cca46afb95660e18b77c583af31

**Comment by kares** [comment]
> When creating a child process from jruby, the child process should not copy the whole memory of the parent process and should probably use copy-on-write mode of forking (vfork instead of fork if I'm not mistaken).

that is not the way external (child) processes work - the JVM pretty much can-not be forked really.

> We are running a large application (8 gb of memory) on a box with 16gb total memory. When our app calls external program (small memory usage), the child process tries to allocate 8 more GBs of memory and therefore crashes with Errno::ENOMEM: Cannot allocate memory.****

after how many calls does it fail?
sounds like there might be some kind of leak - probably externally (outside the JVM).
is it possible to know how exactly are you calling the external program and maybe even what program?

**Comment by AlexEzzeddine** [comment]
> after how many calls does it fail?
> sounds like there might be some kind of leak - probably externally (outside the JVM).
> is it possible to know how exactly are you calling the external program and maybe even what program?

Can you run this program on a 4gb box or container: https://gist.github.com/AlexEzzeddine/5b493cca46afb95660e18b77c583af31 
The command you should run is `JRUBY_OPTS="-J-Xmx3g" jruby test_fork.rb

**Comment by headius** [comment]
I'm looking into this a bit now. It seems we can specify that we want `posix_spawn` to use `vfork` explicitly, but apparently on Linux it will also use `vfork` automatically under some [conditions](http://man7.org/linux/man-pages/man3/posix_spawn.3.html):

```markdown
       The child process is created using vfork(2) instead of fork(2) when
       either of the following is true:

       *  the spawn-flags element of the attributes object pointed to by
          attrp contains the GNU-specific flag POSIX_SPAWN_USEVFORK; or

       *  file_actions is NULL and the spawn-flags element of the attributes
          object pointed to by attrp does not contain
          POSIX_SPAWN_SETSIGMASK, POSIX_SPAWN_SETSIGDEF,
          POSIX_SPAWN_SETSCHEDPARAM, POSIX_SPAWN_SETSCHEDULER,
          POSIX_SPAWN_SETPGROUP, or POSIX_SPAWN_RESETIDS.

       In other words, vfork(2) is used if the caller requests it, or if
       there is no cleanup expected in the child before it exec(3)s the
       requested file.
```

We are not using any of the prohibited attributes, but obviously `popen3` juggles file descriptors around. So it would be using fork in this case.

We can try to define and pass the `POSIX_SPAWN_USEVFORK` when we know we're on GNU libc, I suppose.

**Comment by headius** [comment]
Interestingly it looks like OpenJDK just recently fixed a similar issue: https://bugs.openjdk.java.net/browse/JDK-8213192

**Comment by headius** [comment]
More info indicating that perhaps `vfork` is not such a great idea to always use: https://ewontfix.com/7/

**Comment by AlexEzzeddine** [comment]
@headius it looks like MRI is using `vfork` if multiple conditions are met: https://github.com/ruby/ruby/blob/master/process.c#L3960
For example, they check if the parent process is not privileged to avoid issues described here https://ewontfix.com/7/
Also looks like they are disabling signal handlers in the child process: https://github.com/ruby/ruby/blob/master/process.c#L3971

**Comment by AlexEzzeddine** [comment]
We are currently in the process of upgrading our applications from 1.7 to 9k and our upgrade is blocked by this issue because there is a risk multiple threads might call external process at the same time and fail with OOM error. Is there any chance the fix could be included into 9.2.9.0 ? Is there any way I could help to get this fixed?

**Comment by headius** [comment]
A not great workaround for you would be to disable native library support with `-Xnative.enabled=false` or JVM property `-Djruby.native.enabled=false`, which will avoid the use of our `posix_spawn` logic. Unfortunately it's not possible to *just* disable the `posix_spawn` process logic; that might be a flag we can add for 9.2.9.

We certainly could try to add the `posix_spawn` vfork logic but I am not clear under what circumstances it's truly safe to use. Every resource I find online discourages its use (sometimes in strong terms).

**Comment by headius** [comment]
@AlexEzzeddine See https://github.com/jruby/jruby/pull/5921 for an option that just disables native process logic. If your application works with this, it would be a better workaround for you until we decide how to handle this vfork vs fork issue.

**Comment by headius** [comment]
@enebo I don't think we should try to shoehorn vfork support into 9.2.9, especially if we can land this native popen option I mention above. We need to figure out when and if we can force posix_spawn to use vfork and do some rigorous testing before it goes into a release.

**Comment by enebo** [comment]
@headius yeah seems less risky to me.  We need to bake any change for vfork to make sure we don't cause a bigger problem.

**Comment by AlexEzzeddine** [comment]
> @AlexEzzeddine See #5921 for an option that just disables native process logic. If your application works with this, it would be a better workaround for you until we decide how to handle this vfork vs fork issue.

@headius thanks, I will try this out and will get back to you.

**Comment by AlexEzzeddine** [comment]
@headius Thanks for adding the option in #5921, I just tested it.
When running our application with `jruby.native.enabled=false`, I found that all calls to `minimagick` gem fail with this output (`jruby.native.enabled=true` works fine):
```
/tmp/test.jpg JPEG 1000x1000 1000x1000+0+0 8-bit sRGB 31.4KB 0.000u 0:00.000
identify: unable to open image `{:in=>#<IO:fd 156>, :out=>#<IO:fd 159>, :err=>#<IO:fd 162>}': No such file or directory @ error/blob.c/OpenBlob/2701.
identify: no decode delegate for this image format `' @ error/constitute.c/ReadImage/504.
```
After some debugging, here is what I found:
`minimagick` gem tries to run this command `['identify', '/tmp/test.jpg']` with [popen3](https://github.com/minimagick/minimagick/blob/v4.9.2/lib/mini_magick/shell.rb#L43): 
The method call `minimagick` does is `Open3.popen3('identify', '/tmp/test.jpg')`

As you know, [popen methods add](https://github.com/jruby/jruby/blob/master/lib/ruby/stdlib/open3.rb#L87) `opts` hash with pipes and [pass it as an argument](https://github.com/jruby/jruby/blob/master/lib/ruby/stdlib/open3.rb#L202) to `Process.spawn` method. Looks like `PopenExecutor.spawn` method can properly parse this `opts` hash but `ShellLauncher.runExternalWithoutWait` cannot parse it and passes it to the external process as an argument to external command.

Please see [this gist](https://gist.github.com/AlexEzzeddine/c34abfafc389d5c27e25bb64cf35181d) for a reproducible test case. Also when run with `jruby.native.enabled=false` this gist fails with:
```
warning: thread "Ruby-0-Thread-1: /home/vagrant/.rvm/rubies/jruby-head-alex/lib/ruby/stdlib/open3.rb:203" terminated with exception (report_on_exception is true):
Errno::ECHILD: No child processes - No child processes
```
 Looks like it fails in [Process.detach](https://github.com/jruby/jruby/blob/master/lib/ruby/stdlib/open3.rb#L203), but it doesn't happen with `jruby.native.enabled=true`


Do you think it's possible to make `PopenExecutor.spawn` and `ShellLauncher.runExternalWithoutWait` methods interchangable?

**Comment by headius** [comment]
BTW I'm surprised that this worked on 1.7, because the flag I added to use non-native process logic basically reverts us to 1.7 behavior. Can you confirm why this is? Are there other updates involved?

There's also some "*magick" libraries for JRuby that don't use ImageMagick and instead just use in-process JDK image processing libraries. @enebo might be able to offer some suggestions since he maintains one of them: https://github.com/jruby/image_voodoo

Also, it seems like `posix_spawn` may use vfork in more cases on a more recent glibc. Is this possibly helpful? https://sourceware.org/bugzilla/show_bug.cgi?id=10354

There are other links indicating that some implementations of `posix_spawn` use vfork more aggressively than others.

**Comment by enebo** [comment]
Also see https://github.com/Serabe/RMagick4J.  It should work and if it doesn't we can fix it (although if image_voodoo can do what you want that is definitely preferred).

**Comment by AlexEzzeddine** [comment]
> BTW I'm surprised that this worked on 1.7, because the flag I added to use non-native process logic basically reverts us to 1.7 behavior. Can you confirm why this is? Are there other updates involved?

our 1.7 branch uses `minimagick:3.6.0` but our 9k branch uses `minimagick:4.9.2` which moved from using `subexec` gem to `popen3`. I think it's not important though because both these methods use `Kernel.spawn` method under the hood but in jruby 1.7 `Kernel.spawn` used to parse arguments: https://github.com/jruby/jruby/blob/1_7_20/core/src/main/ruby/jruby/kernel19/process.rb#L13

Maybe we can do the same thing in jruby 9k?

**Comment by AlexEzzeddine** [comment]
> Also, it seems like posix_spawn may use vfork in more cases on a more recent glibc. Is this possibly helpful? https://sourceware.org/bugzilla/show_bug.cgi?id=10354

This is also a good option, we are looking at updating to ubuntu 18.04 which has the latest `libc`.

**Comment by AlexEzzeddine** [comment]
> There's also some "*magick" libraries for JRuby that don't use ImageMagick and instead just use in-process JDK image processing libraries. @enebo might be able to offer some suggestions since he maintains one of them: https://github.com/jruby/image_voodoo

Looking at both `image_voodoo` and `RMagick4J` I wonder if they can integrate with `carrierwave` gem and what the performance differences are across the 3 gems.

**Comment by enebo** [comment]
@AlexEzzeddine Looking at https://github.com/carrierwaveuploader/carrierwave/blob/master/lib/carrierwave/processing/mini_magick.rb I think image_voodoo will probably work out.  I did not fully grok things like resize_to_fill (is it a crop) but I am willing to add anything missing to image_voodoo if it is actually missing.  Most of the stuff there is what image_voodoo is about.  I expect image_voodoo to win performance wise but I have never did a head to head.

**Comment by AlexEzzeddine** [comment]
Confirmed that this issue is not present in Ubuntu 18.04 because of the newer libc version. More details here: https://sourceware.org/bugzilla/show_bug.cgi?id=10354

**Comment by headius** [comment]
@alexdowad That's great to hear!

I have been a bit paralyzed by all the different posts that either encourage or discourage the POSIX_VFORK flag. It seems to me that most folks agree it should only be used if you yourself audit the posix_spawn options to make sure they're vfork-compatible. I think this is a poor and fragile way to go.

My current position (based on your successful result) is that we recommend people update their glibc or ask their C library maintainers to make similar improvements. It should not be up to the callers of posix_spawn to know which options break under vfork in a given environment

What do you think?

**Comment by AlexEzzeddine** [comment]
I think that this issue is definitely a bug (just not sure in which library) because when I spawn a light external process I don't expect it to claim as much memory as the parent process uses.
Also as I mentioned in previous comments, MRI prefers `vfork` over `fork` when some conditions are met: (https://github.com/jruby/jruby/issues/5911#issuecomment-542490834). Isn't jruby supposed to have same behavior as MRI?
But I agree it's too risky to make changes to posix_spawn behavior without fully understanding the consequences and until we fully understand them maybe we can just wait for more feedback and see how many users encounter the same issue. 
My guess is that not many at all since this was the first github issue in this repo ever created on this topic so there is no urgency here. Also for some of those who will encounter this issue we have a solution (upgrading libc).
Thanks!

**Comment by headius** [comment]
@AlexEzzeddine MRI has more flexibility here because they can fork, do some stuff, and then exec later on. Because we are running under the JVM, we are limited to available ways to fork+exec in one shot, which basically means posix_spawn in *nix and CreateProcess in Windows (though we are still relying on the JVM's process launching on Windows).

So we could probably do the same filtering that MRI does and pass the non-standard VFORK flag to posix_spawn, but these "privileged" operations are a known quantity and it would seem a far better option to fix it in posix_spawn.

FWIW I do not see this behavior on MacOS Mojave. Spawning `bash -c 'read'` from JRuby produces a bash process that's under 1MB in size. I assume Windows is unaffected because it doesn't fork. I have not tested other operating systems.

So at this point I'd say we would welcome someone implementing the filtering and VFORK flag but we're not going to spend cycles on it.

**Comment by headius** [comment]
Oh another comment about the filtering: MRI does this filtering knowing that it can do a lot of fiddling around before the eventual `exec` call. A filter for posix_spawn would need to know more about how posix_spawn implements the various options, and as we've seen from doco there's not a lot of consensus on when it is safe to posix_spawn VFORK and when it is not.

**Comment by sachinswagh** [comment]
@headius : Any updates on this? I am using jruby 9.1.17 and facing `ERROR Cannot allocate memory`. I am blocked on this.
